src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -287,7 +287,11 @@ public static final class QueueTaskCancelled extends CauseOfInterruption {
                             continue;
                         }
                         listener.getLogger().println("Agent " + node.getNodeName() + " was deleted; cancelling node body");
-                        body.cancel(new RemovedNodeCause());
+                        if (Util.isOverridden(BodyExecution.class, body.getClass(), "cancel", Throwable.class)) {
+                            body.cancel(new FlowInterruptedException(Result.ABORTED, false, new RemovedNodeCause()));
```

Comment:
Passing `false` here is the key aspect of the fix. It means that this was _not_ some user clicking the red **×** on the build, and so `retry` should process it by default (even without use of `conditions: [agent()]`).